[ti_cif3] Add New TI integration for the Collective Intelligence Framework v3

mdavis332 opened the pull request:
## What does this PR do?

* Added a new TI package integration for [Collective Intelligence Framework v3](https://github.com/csirtgadgets/bearded-avenger/) (CIFv3)
* Added a data stream
* Added data collection logic to the data stream
* Added the ingest pipeline for the data stream
* Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files
* Added dashboards and visualizations
* Added test for pipeline for the data stream
* Added system test cases for the data stream
* Included test results were generated automatically

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

* Clone integrations repo
* Install elastic package locally
* Start elastic stack using `elastic-package stack up -v -d`
* Move to integrations/packages/ti_cif3 directory.
* Run the following command to run tests:
> elastic-package test

## Screenshots
![fleet-add_integration](https://user-images.githubusercontent.com/1747636/181106434-5200cae5-d5e3-4317-be67-714330912b26.jpg)

![kibana-ip_dashboard](https://user-images.githubusercontent.com/1747636/181106451-86ba7129-ebb8-4d02-bf6e-b6c1df3bd2a8.jpg)

![test_asset](https://user-images.githubusercontent.com/1747636/181107227-180a35c1-c959-4bda-b5df-d9430f590112.jpg)

![test_pipeline](https://user-images.githubusercontent.com/1747636/181107230-6120cec3-61f0-4c8b-9ff5-11e3c4da44c2.jpg)

![test_static](https://user-images.githubusercontent.com/1747636/181107232-45e81b81-512d-4d42-a817-4d643afd3bc9.jpg)

![test_system](https://user-images.githubusercontent.com/1747636/181107233-446c5f85-98c7-4921-ae24-3660f67acea3.jpg)

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3839/detail/PR-3839/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3839/detail/PR-3839/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3839/detail/PR-3839/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3839/detail/PR-3839/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-26T20:28:23.046Z&rangeTo=2022-07-26T20:48:23.046Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=80ea623c4eb0d4df1c439e04a1ab3f11&transactionId=4a14d820a169f6c1)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-26T20:38:23.046+0000


* Duration: 3 min 47 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3839/detail/PR-3839/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a shared library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3839/runs/1/steps/81/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/integrations.yml</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3839/runs/1/steps/117/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3839/runs/1/steps/89/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/mdavis332 return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/mdavis332 : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/mdavis332 : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/reference/orgs#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mdavis332 commented:
/test

jsoriano commented:
Hey @mdavis332, as mentioned in https://github.com/elastic/integrations/pull/3904#issuecomment-1199399786, please check that you are using the same email address in your commits that you used to sign the CLA.

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

peasead commented:
What I really like about this is the inclusion of Threat Fox as a Provider. It's a high-quality and contextual feed.

peasead commented on packages/ti_cif3/data_stream/feed/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,341 @@
+      field: threat.indicator.type
+      value: file
+      if: "['md5', 'sha1', 'sha256', 'sha512', 'ssdeep'].contains(ctx.cif3?.itype) && !ctx.cif3?.tags.contains('ja3')"
+  - rename:
+      field: cif3.indicator
+      target_field: tls.client.ja3
+      ignore_missing: true
+      if: "ctx.cif3?.itype == 'md5' && ctx.cif3?.tags.contains('ja3')"
```
Looking at [field reuse](https://www.elastic.co/guide/en/ecs/current/ecs-threat.html#_field_reuse_24) for the Threat ECS fieldset, I think this is a gap that needs to be addressed by adding `tls.` as a valid reuse field for `threat.*`.

Ideally, we'd want this to be something like `threat.indicator.tls.client.ja3`. 

@ebeahan would we be able to make a parallel effort in adding TLS as a valid reuse field and adjusting this pipeline to be `threat.indicator.tls.client.ja3` or do we need to adjust ECS first?

mdavis332 replied:
@peasead thanks so much for carefully looking at the integration. Although I'm not elastic staff, I agree with the suggestion of adding the TLS field for reuse under `threat.indicator`. I had the same thought about `threat.indicator.network.cidr`. Although it's not currently on the main [ECS-threat reuse section](https://www.elastic.co/guide/en/ecs/current/ecs-threat.html#_field_reuse_24), I went ahead and used that `network.cidr` field under `threat.indicator` due to having read through this related convo: https://github.com/elastic/beats/issues/29949

peasead replied:
Sure thing.

This all looks really good and we'll wait from @ebeahan to see what his thoughts are on the ECS side.

ebeahan replied:
Apologies - missed the mention here until now.

No objection on the naming for `threat.indicator.tls.client.ja3`. Since the field doesn't exist in ECS, it will need defining in the integration itself.

If there's interest to add `tls.*` field reuse under `threat.indicator.*`, I think it's a small enough addition it can be suggested in a PR vs. through an ECS RFC.

mdavis332 replied:
Hi @ebeahan,

> Since the field doesn't exist in ECS, it will need defining in the integration itself.

Would that involve tweaking [data_stream/feed/fields/ecs.yml](https://github.com/elastic/integrations/pull/3839/files#diff-3a2beaf17553db35ce0c2a7376e5c380dc79d1ac0056b7938a6ac216269bad0e)
```
  - external: ecs
    name: tls.client.ja3
```
to
```
  - external: ecs
    name: threat.indicator.tls.client.ja3
```
?

> I think it's a small enough addition it can be suggested in a PR 

Is that a suggestion for me or elastic staff? I'm not as familiar with https://github.com/elastic/ecs. I assume it would need to be added somewhere in https://github.com/elastic/ecs/blob/main/schemas/threat.yml, but I'm unsure.

efd6 replied:
It would need to be an explicit definition, analogous to this https://github.com/elastic/integrations/blob/ce52809ff16c61dae1a33627a897ed1121380ec6/packages/google_workspace/data_stream/groups/fields/ecs.yml#L65-L68

This was done for `geo.location` before it was included in ECS formally as a `geo_point` type.

What needs to be added is just the type and the relevant details for the ja3 field from https://www.elastic.co/guide/en/ecs/current/ecs-tls.html#field-tls-client-ja3.

mdavis332 replied:
Thanks @efd6. I just commited b6e54db63ad55ae8841547f2c1bb8587a99274d2 which hopefully resolves that suggestion.

mdavis332 commented:
> What I really like about this is the inclusion of Threat Fox as a Provider. It's a high-quality and contextual feed.

just so there's no unmet expectations, ThreatFox isn't inherently part of this integration or one of the default CIFv3 ingestion rules, but since it's public info and a worthwhile provider of info, I included it as our pipeline test example :) It's definitely possible to configure as a an ingestion rule in [csirtg-smrt](https://github.com/csirtgadgets/csirtg-smrt-v1/), the ingest module of CIFv3.

eriroley commented:
Because I'm looking to use this - what is the current status of this integration? and what still needs to happen?
I see "integrations/pr-merge — The build of this commit was aborted" with "The PR is not allowed to run in the CI yet"

mdavis332 commented:
> Because I'm looking to use this - what is the current status of this integration? and what still needs to happen? I see "integrations/pr-merge — The build of this commit was aborted" with "The PR is not allowed to run in the CI yet"

hey @eriroley, I think we may be waiting on @ebeahan to give his thoughts on the ECS side of things. Not sure what else may be required.

mdavis332 commented:
To anyone from the elastic team: just checking in to see if there's anything I can do to help move forward here.

peasead commented:
Checking internally on the status.

mdavis332 commented:
> Checking internally on the status.

Hi @peasead, just my weekly checkin :)

There's an event coming up in the middle of September for which I'd love this to be ready to install by end users. If it's not possible to make it into the EPM by then, is there an easy way for end-users to download the integration and locally load it? Kinda like installing/side-loading a local plugin from zip or something?

Thanks for checking, and nice work on the recent blog post for the elastic container project.

jamiehynds commented:
@mdavis332 apologies for the delay on getting your PR reviewed. We're working through a backlog of new integration PR's at the moment, so it's taking a little longer than expected. We'll certainly aim to have it published in time for your mid-September event. 

@P1llus do you have the bandwidth to review this PR over the next week or so? No worries if not, can ask someone else on the team to review.

efd6 commented on packages/ti_cif3/data_stream/feed/_dev/test/pipeline/test-common-config.yml:
```diff
@@ -0,0 +1,5 @@
+fields:
+  tags:
+    - preserve_original_event
+dynamic_fields:
+  event.ingested: "^.*$"
```
This is not required.

mdavis332 replied:
Removed in commit 4e6fe712f87e5c303e1b305d27a78b1d8a52ce6d

efd6 commented on packages/ti_cif3/data_stream/feed/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,341 @@
+  - grok:
+      field: cif3.indicator
+      patterns:
+        - "%{URIPROTO:threat.indicator.url.scheme}://(?:%{IP:threat.indicator.ip}|%{HOSTNAME:threat.indicator.url.domain})(?::%{POSINT:threat.indicator.url.port})?(?:%{URIPATH:threat.indicator.url.path})?.*"
```
Are we safe with IPv6/port values here?

mdavis332 replied:
Good callout. Indeed, ipv6/port values are good here, but you did help me realize that the IPV6 convention of putting square brackets around the address if a port is involved wouldn't have previously matched. I've updated the pattern in commit 478bb34d4c2063800391bb24d3f7bd53ce219298.

efd6 replied:
The IPv6/port world is not a happy one. This will probably be OK, but it might be worth taking a look at https://github.com/elastic/integrations/pull/3117#discussion_r853610542.

mdavis332 replied:
Thanks for that reference. I couldn't find any references to IPV6NOCOMPRESS or IPV6PORTSEP grok patterns so perhaps those are ES specific? At any rate, URLs with IPv6 addresses from CIF should always have been normalized to look like:
* https://2001:d3ad:b33f:0000:0000:d3ad:b33f:0001/testing OR
* https://[2001:d3ad:b33f:0000:0000:d3ad:b33f:001]:8080/testing if it includes a port

So for this plugin, as long as we catch those cases, I think we _should_ be okay for the moment (famous last words).

efd6 replied:
Sorry, those are defined locally e.g. https://github.com/elastic/integrations/blob/c56c28069734ce6000255eb9efecffab198b633b/packages/cisco_duo/data_stream/auth/elasticsearch/ingest_pipeline/default.yml#L95-L107

efd6 commented on packages/ti_cif3/data_stream/feed/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,341 @@
+      if: ctx.cif3?.confidence != null
+      description: Normalize confidence level.
+      source: >
+        def value = ctx.cif3.confidence;
+        if (value < 0.0 || value > 10.0) {
+          ctx.threat.indicator.confidence = "None";
+          return;
+        }
+        if (value >= 0.0 && value < 3.0) {
+          ctx.threat.indicator.confidence = "Low";
+          return;
+        }
+        if (value >= 3.0 && value < 7.0) {
+          ctx.threat.indicator.confidence = "Med";
+          return;
+        }
+        if (value >= 7.0 && value <= 10.0) {
+          ctx.threat.indicator.confidence = "High";
+          return;
+        }
```
The mapping between the cif3 confidence and the ECS keywords would be helpful in the README.

mdavis332 replied:
Added README info in commit 696c2a05c24d0868dd5e6e41207dae7961ec0123

efd6 commented on packages/ti_cif3/data_stream/feed/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,341 @@
+  ####################
+  - set:
+      field: ecs.version
+      value: "8.3.0"
```
8.4.0 now.

mdavis332 replied:
updated to 8.4.0

mdavis332 commented on packages/ti_cif3/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,13 @@
+# Collective Intelligence Framework v3 Integration
+
+This integration connects with the [REST API from the running CIFv3 instance](https://github.com/csirtgadgets/bearded-avenger/blob/master/cif/httpd/views/feed/__init__.py) to retrieve indicators.
```
updated

efd6 replied:
Thanks, do you know of any docs for the field semantics? The docs in the wiki list some, but not all the fields that are used in the tests. As an aside, can you make a note of the provenance of the test cases (if they are hand crafted, that's fine, but knowing this is helpful in the future).

mdavis332 replied:
I'm not aware of any other public API docs than the one referenced here. The test cases in this PR were generated by actual tests, except for any references to `8.3.0` which I just manually changed to `8.4.0` for the ECS version change.

mdavis332 replied:
> As an aside, can you make a note in the PR description of the provenance of the test cases

Do you mean edit the initial PR comment to include that the test results were generated by automated testing? I've looked at other PRs for new features and haven't found similar documentation so just want to be sure I fulfill the request correctly.

efd6 replied:
Yeah, that's what I mean. And you are right, provenance is rarely provided.

mdavis332 replied:
Ok, added a final bullet to the initial PR comment.

efd6 replied:
Thanks

efd6 replied:
Sorry, I was thinking of the input ([test-cif3-sample.ndjson.log](https://github.com/elastic/integrations/pull/3839/files#diff-cbbf6d50f16ff4f9040312b539769a0dd7a7c6326cb362ee21fd516e8ac1c0de)) rather than the generated output. If this was hand crafted, that's fine, otherwise, knowing where it came from is helpful.

mdavis332 replied:
It's been several weeks since I put it together, but I'm fairly certain that was a real TLP:WHITE indicator from ThreatFox that came out of a CIFv3 instance after it was ingested there. I may have changed an IP address to comply w/ the elastic requirements for using certain test IPs or something if I recall correctly.

mdavis332 commented on packages/ti_cif3/img/csg_logo_big.svg:
```diff
@@ -0,0 +1,23 @@
+
+<g transform="translate(0.000000,748.000000) scale(0.100000,-0.100000)"
+fill="#000000" stroke="none">
+<path d="M3585 7465 c-855 -88 -1618 -441 -2219 -1025 -659 -640 -1040 -1452
```
I kinda liked the look without the 'G' but maybe the logo should be intact :) good catch, fixed in commit c66d43bc9bd6f242e1c89792a2236dac0bfe4eb6

mdavis332 commented:
Hi @efd6, thanks so much for your time doing a thorough review. I believe I've addressed all of your review feedback and left comments on each with referenced commits. 

Please let me know if there's anything else to adjust at this time. Thank you again!

efd6 commented:
/test

efd6 commented:
The README.md will need to be rebuilt.

mdavis332 commented:
> The README.md will need to be rebuilt.

Done via commit 57a110a5fa2338f4e89c3bb336b9a6f3925eeb92.

efd6 commented:
/test

efd6 commented on packages/ti_cif3/data_stream/feed/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,341 @@
+  - grok:
+      field: cif3.indicator
+      patterns:
+        - "%{URIPROTO:threat.indicator.url.scheme}://(?:%{IPV4:threat.indicator.ip}|\[?%{IPV6:threat.indicator.ip}\]?|%{HOSTNAME:threat.indicator.url.domain})(?::%{POSINT:threat.indicator.url.port})?(?:%{URIPATH:threat.indicator.url.path})?.*"
```
```suggestion
        - "%{URIPROTO:threat.indicator.url.scheme}://(?:%{IPV4:threat.indicator.ip}|\\[?%{IPV6:threat.indicator.ip}\\]?|%{HOSTNAME:threat.indicator.url.domain})(?::%{POSINT:threat.indicator.url.port})?(?:%{URIPATH:threat.indicator.url.path})?.*"
```

efd6 commented:
/test

efd6 commented on packages/ti_cif3/data_stream/feed/_dev/test/pipeline/test-cif3-sample.ndjson.log:
```diff
@@ -0,0 +1 @@
+{"indicator":"89.160.20.156","itype":"ipv4","tlp":"white","provider":"threatfox.abuse.ch","group":["everyone"],"count":1,"tags":["agenttesla","botnet","hunter"],"confidence":8.0,"description":"agent tesla","uuid":"3fbdd654-b2b0-498c-8e20-ef87bce73672","reference":"https://threatfox.abuse.ch/ioc/838651/","rdata":"http://208.67.106.111/theme/inc/e26dbe0dcc481e.php","firsttime":"2022-07-19T07:40:41. ...
```
This file and the expects will need to be renamed to test-cif3-sample-ndjson.log and test-cif3-sample-ndjson.log-expected.json. (You can run `elastic-package check` to check the outcome).

efd6 commented:
/test

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`1/1`) | :green_heart: 2.809
Classes | 100.0% (`1/1`) | :green_heart: 2.809
Methods | 85.714% (`12/14`) | :-1: -3.685
Lines | 39.615% (`103/260`) | :-1: -51.333
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

efd6 reviewed (approved):
LGTM, but please wait for @ebeahan for comment on the ECS issue.

efd6 commented:
Thanks

mdavis332 commented:
> LGTM, but please wait for @ebeahan for comment on the ECS issue.

Thanks again, Dan, for all your time reviewing and working through the CI stuff.

mdavis332 commented:
Hi @jamiehynds, we're coming up on that mid-September timeframe, so just wanted to check back in this week. I appreciate everyone that's contributed to moving this PR forward.

efd6 commented on packages/ti_cif3/data_stream/feed/fields/ecs.yml:
```diff
@@ -58,8 +58,10 @@
-- external: ecs
-  name: tls.client.ja3
+- name: threat.indicator.tls.client.ja3
+  level: core
```
From what I can see, `tls.client.ja3` is [extended](https://www.elastic.co/guide/en/ecs/current/ecs-tls.html#field-tls-client-ja3), but please wait for confirmation from another before changing it.

mdavis332 replied:
Would that be @ebeahan who can confirm? Just trying to quickly knock out all the requests of me so we can get this closed out.

efd6 replied:
I'm fairly confident, so maybe just make the change. If I'm wrong, we can fix it in a follow-up.

mdavis332 replied:
Ok, adjusted in commit b557362f5023d6e2126e619fa1cf8c4d2ddbb184

jamiehynds commented:
@efd6 is this PR ok to merge now? The only outstanding item was the TI ECS query, which Eric commented on and @mdavis332 has addressed. Thanks!

efd6 commented:
🚀 Thanks @mdavis332
